### PR TITLE
crypto: verify aggregated sig with fallback

### PR DESCRIPTION
**Pull request by joyqvq**
## Description 
currently:
1. a narwhal certificate is formed if the vote_aggregator accumulated enought votes (each vote is individually verified), we actually dont verify the aggregated sig at all. 
2. only when a node receives a certificate from another peer, its aggregated sig is verified, and if its invalid, discard the cert. 

this change optimizes (1) to remove individual vote verify and replace it with only verify if the quorum threshold is reached. in the happy path, it is done. otherwise, if there are bad sig found, remove from weights and votes list, then continue to find new vote. (2) is unchanged. 

## Test Plan 

WIP

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: narwhal/primary/src/aggregators.rs
Comment (joyqvq) [inline narwhal/primary/src/aggregators.rs:73]: i cannot directly uses https://github.com/MystenLabs/sui/blob/776c90b88c87d86102bd31b41c320025c2d0de39/narwhal/types/src/primary.rs#L640 bc there is no worker_cache here - is that something we need to verify?

File: narwhal/primary/src/aggregators.rs
Comment (joyqvq) [inline narwhal/primary/src/aggregators.rs:93]: https://github.com/MystenLabs/sui/blob/b21e75596795b328cbaff39be244ab5be23bf3dd/narwhal/primary/src/certifier.rs#L285-L299 returning a None here will have the loop continue to find new votes.

File: narwhal/types/src/primary.rs
Comment (joyqvq) [inline narwhal/types/src/primary.rs:423]: perhaps we want to rename this function to validate since it doesnt actually run verify sig? there are few other things it validates though

**Review by benr-ml — APPROVED** [review]
Nice! The crypto side looks good to me, and also the rest but let's wait for @akichidis approval as well

**Review by benr-ml — COMMENTED** [review]
Oh, let's add a test with an invalid vote to make sure we don't break the assumption regarding the loop

File: narwhal/primary/src/aggregators.rs
Comment (mwtian) [reply narwhal/primary/src/aggregators.rs:73]: Verifying information outside of votes should be unnecessary here, since Header is produced locally.

File: narwhal/types/src/primary.rs
Comment (mwtian) [inline narwhal/types/src/primary.rs:1]: I believe we should also rename `Certificate::new()` into `Certificate::new_unverified()` to make its purpose clear. The vote filtering logic from L565 ~ L580 can also be moved out to the vote aggregator. Alternatively, aggregation signature verification should be done inside `Certificate::new()`, and the aggregator would check its returned error to decide if it needs to fallback.

File: narwhal/primary/src/aggregators.rs
Comment (mwtian) [inline narwhal/primary/src/aggregators.rs:75]: Can we log the error as well or is it not very informative?

**Review by akichidis — COMMENTED** [review]
Thanks @joyqvq for this improvement! The changes make sense to me - waiting for the final version of it

File: narwhal/primary/src/aggregators.rs
Comment (joyqvq) [reply narwhal/primary/src/aggregators.rs:75]: its most likely to be an opaque "invalid signature error" here, but since it can potentially have invalid length etc, we can log it

File: narwhal/types/src/primary.rs
Comment (joyqvq) [reply narwhal/types/src/primary.rs:1]: but i think the vote filtering logic is still needed in certificate::new to construct the signed_authorities bitmap?

File: narwhal/primary/src/aggregators.rs
Comment (benr-ml) [inline narwhal/primary/src/aggregators.rs:80]: not sure how inefficient it is in practice, but note that this function clones each vote 3 times, which could potentially be expensive.

Comment (joyqvq) [reply narwhal/primary/src/aggregators.rs:80]: good point, revised to remove votes in place instead.

File: narwhal/primary/src/tests/certifier_tests.rs
Comment (akichidis) [inline narwhal/primary/src/tests/certifier_tests.rs:363]: You could also add a timeout here to ensure that test will return early if no certificate gets created within a reasonable amount of time

**Review by akichidis — APPROVED** [review]
Thanks @joyqvq for this, great work!

**Pull request merged**
